src/Symfony/Component/Form/Extension/Core/DataMapper/PropertyPathMapper.php

```diff
@@ -76,16 +88,37 @@ public function mapFormsToData($forms, &$data)
                 $propertyValue = $form->getData();
                 // If the field is of type DateTimeInterface and the data is the same skip the update to
                 // keep the original object hash
-                if ($propertyValue instanceof \DateTimeInterface && $propertyValue == $this->propertyAccessor->getValue($data, $propertyPath)) {
+                if ($propertyValue instanceof \DateTimeInterface && $propertyValue == $this->getPropertyValue($data, $propertyPath)) {
                     continue;
                 }
 
                 // If the data is identical to the value in $data, we are
                 // dealing with a reference
-                if (!\is_object($data) || !$config->getByReference() || $propertyValue !== $this->propertyAccessor->getValue($data, $propertyPath)) {
+                if (!\is_object($data) || !$config->getByReference() || $propertyValue !== $this->getPropertyValue($data, $propertyPath)) {
                     $this->propertyAccessor->setValue($data, $propertyPath, $propertyValue);
                 }
             }
         }
     }
+
+    /**
+     * Get the property value per PropertyAccessor.
+     * Treat uninitialized properties as null.
+     *
+     * @param object|array                 $objectOrArray The object or array to traverse
+     * @param string|PropertyPathInterface $propertyPath  The property path to read
+     *
+     * @return mixed The value at the end of the property path
+     *
+     * @throws Exception\InvalidArgumentException If the property path is invalid
+     * @throws Exception\UnexpectedTypeException  If a value within the path is neither object nor array
+     */
+    private function getPropertyValue($data, $propertyPath)
+    {
+        try {
+            return $this->propertyAccessor->getValue($data, $propertyPath);
+        } catch (AccessException $e) {
```

Comment:
Shouldn't we apply the same logic as above to not blindly silence all exceptions that could be thrown here?

Reply:
I don't think there should be any exception for failed reading of a property when trying to write it.

Reply:
If there cannot be any exceptions, we shouldn't add the `catch` here to not hide any bugs IMO.

Reply:
The problem is that the implementation tries to check for the current value before calling setValue.

```php
class Foo1 {
    public int $bar;
}
```
One should be able to map a form with `bar=123` to `Foo1`, but it will fail with `UninitializedPropertyException`.

```php
class Foo2 {
    private $bar;
    public function setBaz($value) {
        $this->bar = $value * 2;
    }
}
```
One should be able to map a form with `baz=123` to `Foo2`, but it will fail with a `NoSuchPropertyException`.


It is at least confusing and IMO a bug, that setting a value fails because reading it is not possible.

I decided against catching `Throwable` here, because e.g. `RuntimeException` thrown by `PropertyAccessor` points to bugs in the implementation with which one can not expect the consecutive `setValue()` to work.

Maybe @HeahDude can give further insight, but as I understood the current implementation the `AccessException` types point to errors in a specific read or write acess, while the other exeptions point to wrong usage of `PropertyAccessor`.

Reply:
I am not against ignoring errors caused by uninitialized properties. I just think that ignoring any other errors that might happen is not the desired solution. I suggest to update the new method like this:

```php
private function getPropertyValue($data, $propertyPath)
{
    try {
        return $this->propertyAccessor->getValue($data, $propertyPath);
    } catch ( $exception) {
        // Skip unitialized properties on $data
        // For versions without UninitializedPropertyException check the exception message
        if (!$e instanceof UninitializedPropertyException && (class_exists(UninitializedPropertyException::class) || false === strpos($e->getMessage(), 'You should initialize it'))) {
            throw $e;
        }
    }
}
```

We can then reuse this method in the `setData()` call above like this:

```php
$form->setData($this->getPropertyValue($data, $propertyPath));
```

Reply:
When mapping form to data we try to *set* values on the object. This should work for properties that can be written on, but not read from, the object - also if the reading fails for other reasons than an unitialized property. It should only fail if writing such property is not possible.


When mapping data to form we try to *get* values from the object. This should not set anything for properties that are not initialized - especially not set null for those which are not nullable typed.
But there might be other reasons for it to fail when reading such property.
```php
class Foo {
 public string $bar;
 public string $baz;
}

// this should be the same
$foo = new Foo();
$foo->bar = 'bar';
$propertyPathMapper->mapDataToForm($foo, [$someForm]);

// as this
$propertyPathMapper->mapDataToForm(['bar' => 'bar'], [$someForm]);
```

Reply:
> When mapping form to data we try to set values on the object. This should work for properties that can be written on, but not read from, the object - also if the reading fails for other reasons than an unitialized property. It should only fail if writing such property is not possible.

I don't agree here. A form requires that data can be read and written. We shouldn't hide any misconfiguration here as that will make debugging a lot harder if the form silently doesn't behave as expected (see also https://github.com/symfony/symfony/pull/36492#discussion_r410931246).

Reply:
It does not silently discard relevant exceptions here. When one tries to write data, one expects it to fail if it can't write it or can not guarantee a certain result.
But here it refuses to write, because it can not guarantee that it doesn't overwrite a property with an identical value. So worst case you end up with what you wanted to be there being there.

If the writing fails due to any reason an exception will be thrown on the next line.

Please consider the example `Foo2` from https://github.com/symfony/symfony/pull/36492#discussion_r416474125 again.

Imagine a form that is input-only and discarded after the object is created.
How is it easier to debug that mapping `baz` to `Foo2::$baz` per `Foo2::setBaz()` fails, because `Foo2` has no `::getBaz()` method?

And the fix is to implement `public function getBaz():void {}`. A method that doesn't do anything beside taming the PropertyPathMapper.

For me that is counterintuitive and required a thorough reading of both the FormMapping and PropertyAccessor to understand why I got an exception for a failed read, when I only tried to write.

Reply:
> And the fix is to implement `public function getBaz():void {}`. A method that doesn't do anything beside taming the PropertyPathMapper.

That's true, but the Form component has always have that assumption as it wasn't designed for write-only forms from the beginning. Changing that would be a new feature, but nothing that we merge in a patch release.

Reply:
I still don't see any harm done. What worked would still work, and what should fail would fail.
The current behavior is neither documented nor to be expected.

But to move this foward: Will this get an approval, if I change and comment the code here and make the full request again for the master?

Reply:
> But to move this foward: Will this get an approval, if I change and comment the code here and make the full request again for the master?

I don't know. I guess I would rather open an issue first and gather some feedback.

Reply:
@ph-fritsche Do you still like to finish the PR though?

Reply:
Yes, of course. I was just waiting for the feedback in https://github.com/symfony/symfony/issues/36754.

> We can then reuse this method in the `setData()` call above like this:
> ```
> $form->setData($this->getPropertyValue($data, $propertyPath));
> ```

That would replace the old bug by a new one as it sets data that is not there.

> ```
> class Foo {
>  public string $bar;
>  public string $baz;
> }
> 
> // this should be the same
> $foo = new Foo();
> $foo->bar = 'bar';
> $propertyPathMapper->mapDataToForm($foo, [$someForm]);
> 
> // as this
> $propertyPathMapper->mapDataToForm(['bar' => 'bar'], [$someForm]);
> ```

Reply:
I am afraid I do not understand. What exactly would be the new bug you are talking about?

As far as I can see #36754 would be a new feature affecting `master` only. So I do not really see how that would affect this PR.

Reply:
The Mapper should transfer data from one representation to the other.
It should fail if it either can not do its job or if it can not guarantee a certain behavior.

The previous implementation introduced restrictions that are neither documented nor are intended. (They could not be, as the pivotal problem leading to this PR came through type hinted properties that didn't exist when the implementation was written.)

Now you could argue that supporting those new concepts is by itself a new feature - then one should close this PR and constrain the dependency to `PHP<7.4`.
If one supports those, one should do it right. And that means that
```php
class Foo {
  // this
  public string $bar;
  // is different than this
  public string $baz = null;
}
// and this should never map null to someForm['bar']
$propertyPathMapper->mapDataToForm($foo, [$someForm]);
```

I committed another change that disables using `mapFormToData()` with setters when the getter is missing, as you considered this a new feature.
(Funny side note: `mapDataToForm()` works, if the setter is missing.)